os/kernel/log_dump/log_dump.c

```diff
@@ -0,0 +1,172 @@
+
+void log_dump_init(void) {
+	sq_init(&log_dump_chunks);
+}
+
+int add_char_to_log_net(char ch)
+{
+	if (sq_empty(&log_dump_chunks)) {
+		/* no chunks are present so we allocate new chunk */
+		struct log_net_chunk_s *node = (struct log_net_chunk_s *)kmm_malloc(sizeof(struct log_net_chunk_s));
+
+		if (node == NULL) {
+			return LOG_DUMP_MEM_FAIL;
+		}
+
+		curbytes = 1;
+		node->arr[0] = ch;
+		sq_addlast((sq_entry_t *)node, &log_dump_chunks);
+		/* set the size of chunks to 1 chunk */
+		log_net_size = sizeof(struct log_net_chunk_s);
+	} else {
+		/* need to check if the current chunks size is over max_log_size or greater than x% of free heap */
+		struct mm_heap_s *heap = mm_get_heap(sq_peek(&log_dump_chunks)); /* get heap details using the head of the chunks linked list */
+
+		if (heap == NULL) {
+			return LOG_DUMP_HEAP_FAIL;
+		}
+
+		size_t free_size = heap->mm_heapsize - heap->total_alloc_size; /* get the free size */
+		size_t percent_free_size = (free_size * CONFIG_LOG_DUMP_MAX_FREE_HEAP) / 100;
+		size_t max_size = MIN(percent_free_size, CONFIG_LOG_DUMP_MAX_SIZE);
+
+		while (!sq_empty(&log_dump_chunks) && log_net_size > max_size) {
+			struct log_net_chunk_s *node = (struct log_net_chunk_s *)sq_remfirst(&log_dump_chunks);
+			kmm_free(node);
+			log_net_size -= sizeof(struct log_net_chunk_s);
+		}
+
+		struct log_net_chunk_s *log_net_tail = (struct log_net_chunk_s *)sq_tail(&log_dump_chunks);
+
+		if (log_net_tail == NULL || curbytes == CONFIG_LOG_DUMP_CHUNK_SIZE) { /* all chunks were removed or last chunk is full */
+
+			/* there is memory, so fetch a new buffer */
+			struct log_net_chunk_s *node = (struct log_net_chunk_s *)kmm_malloc(sizeof(struct log_net_chunk_s));
+
+			if (node == NULL) {
+				return LOG_DUMP_MEM_FAIL;
+			}
+
+			curbytes = 1;
+			node->arr[0] = ch;
+			sq_addlast((sq_entry_t *)node, &log_dump_chunks);
+			log_net_size += sizeof(struct log_net_chunk_s);
+
+		} else { /* there is place in tail */
+
+			log_net_tail->arr[curbytes] = ch;
+			curbytes += 1;
+		}
+	}
+	return LOG_DUMP_OK;
+}
+
+ssize_t read_log_net(FAR char *buffer, size_t buflen)
+{
+	ssize_t ret = 0;
+	int i, k;
+
+	if (!sq_empty(&log_dump_chunks)) { /* check if there are any logs saved */
+
+		if (buflen % CONFIG_LOG_DUMP_CHUNK_SIZE == 0) {
+			int num_chunks = buflen / CONFIG_LOG_DUMP_CHUNK_SIZE;
+			struct log_net_chunk_s *next_chunk;
+			for (i = 0; i < num_chunks; i++) {
+				next_chunk = (struct log_net_chunk_s *)sq_remfirst(&log_dump_chunks);
+				if (sq_empty(&log_dump_chunks)) { /* last chunk */
+					for (k = 0; k < curbytes; k++) {
+						buffer[ret] = next_chunk->arr[k];
+						ret++;
+					}
+					kmm_free(next_chunk);
+					break;
+				} else {
+					for (k = 0; k < CONFIG_LOG_DUMP_CHUNK_SIZE; k++) {
+						buffer[ret] = next_chunk->arr[k];
+						ret++;
+					}
+					kmm_free(next_chunk);
+				}
+			}
+		} else {
+			/* support only read sizes that are multiples of chunk size */
+			ret = LOG_DUMP_SIZE_FAIL;
```

Comment:
This is difficult to use. How about making a new API to give the log size saved?

Reply:
Hi @sunghan-chang , could you please elaborate on this. Currently, only multiples of chunk sizes are allowed while reading. Should I add an api to get the read size?

Reply:
We can check how many node is linked. Then you can give the size of saved log.
In user point of view, I think it's much difficult that how many size should allocated, how many times should get, how to know "multiples of chunk size" even user is not log saving module.

Reply:
I will add support for flexible read sizes instead of supporting only chunk sizes. Also, I think we need not provide the user the api for log size as he can read until read_log returns the size read as 0.

Reply:
Ok. Then could you make the example as you said? Currently it does not distinguish `read done` case and `read failure` case like this line.

Reply:
If there is no error return after applying flexible read size, then no need to change the example.